core/tree/multi_root_tree.go

```diff
@@ -61,36 +65,74 @@ func (t *MultiRootTree) IsCyclic() error {
 	return nil
 }
 
+// get sorted dataMap keys
+func (t *MultiRootTree) getSortedKeys() []string {
+	keys := []string{}
+	for k := range t.dataMap {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+func (t *MultiRootTree) isCyclic(root *TreeNode, visited map[string]bool) error {
+	pathMap := make(map[string]bool)
+	paths := []string{}
+	return t.hasCycle(root, visited, pathMap, &paths)
+}
+
 // runs a DFS on a given tree using visitor pattern
-func (t *MultiRootTree) hasCycle(root *TreeNode, visited, pathMap map[string]bool) error {
+func (t *MultiRootTree) hasCycle(root *TreeNode, visited, pathMap map[string]bool, paths *[]string) error {
 	_, isNodeVisited := visited[root.GetName()]
 	if !isNodeVisited || !visited[root.GetName()] {
 		pathMap[root.GetName()] = true
 		visited[root.GetName()] = true
+		*paths = append(*paths, root.GetName())
 		var cyclicErr error
 		for _, child := range root.Dependents {
 			n, _ := t.GetNodeByName(child.GetName())
 			_, isChildVisited := visited[child.GetName()]
 			if !isChildVisited || !visited[child.GetName()] {
-				cyclicErr = t.hasCycle(n, visited, pathMap)
+				cyclicErr = t.hasCycle(n, visited, pathMap, paths)
 			}
 			if cyclicErr != nil {
 				return cyclicErr
 			}
 
 			_, childAlreadyInPath := pathMap[child.GetName()] // 1 -> 2 -> 1
 			if childAlreadyInPath && pathMap[child.GetName()] {
-				cyclicErr = fmt.Errorf("%w: %s", ErrCyclicDependencyEncountered, root.GetName())
+				*paths = append(*paths, child.GetName())
+				cyclicErr = fmt.Errorf("%w: %s", ErrCyclicDependencyEncountered, stringifyPaths(*paths))
```

Comment:
not sure if `stringify` term is correct for this as it is already an array of strings. should we call it `joinPaths` or you have another suggestion?

Reply:
`printPaths`? or maybe `prettifyPaths`? As it's printing the array paths in tree fashion(?)